### PR TITLE
SQS Plumbing

### PR DESCRIPTION
**Pull request by mike12345567**
## Description
This PR introduces the plumbing for SQS support.

**Please note - by default this does not change anything and SQS implementation is not completed.**

The purpose of this is mostly moving around a few components of the SQL sub-system to try and make it easier to re-use for SQS, as this changes/moves things around quite a bit I've ran into a lot of issues with the plumbing conflicting. I'd like to get this body of work in, so that tests and final implementation can be implemented on top of this, without needing to fix conflicts regularly.

This is backwards compatible, with a feature flag `SQS_SEARCH_ENABLE` which can be set to switch to the SQS search mode - this currently does not handle table design doc creation yet/migrations, so generally enabling this flag will break internal tables search functionality.

A lot of this work is trying to move typing around/improve it to make sharing the internal search process with a SQL search more accurate. The typings around for example pagination weren't exactly accurate and this attempts to improve on a lot of this.

I've added a flag for now that checks if running in production it will error and not start the service which this flag enabled.

### CODE REVIEW

**Review by samwho — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -73,7 +73,7 @@
     "build:docker:single:multiarch": "docker buildx build --platform linux/arm64,linux/amd64 -f hosting/single/Dockerfile -t budibase:latest .",
     "build:docker:single": "./scripts/build-single-image.sh",
     "build:docker:dependencies": "docker build -f hosting/dependencies/Dockerfile -t budibase/dependencies:latest ./hosting",
-    "publish:docker:couch": "docker buildx build --platform linux/arm64,linux/amd64 -f hosting/couchdb/Dockerfile -t budibase/couchdb:latest -t budibase/couchdb:v3.2.1 --push ./hosting/couchdb",
+    "publish:docker:couch": "docker buildx build --platform linux/arm64,linux/amd64 -f hosting/couchdb/Dockerfile -t budibase/couchdb:v3.2.1-sqs --push ./hosting/couchdb",
```
[the remark this answers is not in the document]
Comment (mike12345567) [reply package.json:76]: Done - added a new SQS Dockerfile (`.v2`).

File: packages/backend-core/src/db/couch/utils.ts [hunk 2]
```diff
@@ -30,8 +30,13 @@ export async function directCouchUrlCall({
     },
   }
   if (body && method !== "GET") {
-    params.body = JSON.stringify(body)
-    params.headers["Content-Type"] = "application/json"
+    if (typeof body === "string") {
```
[the remark this answers is not in the document]
Comment (mike12345567) [reply packages/backend-core/src/db/couch/utils.ts:33]: Discussed on call - needed for calls through to the new SQS service, some errors return as string bodies.

File: packages/backend-core/src/db/lucene.ts [hunk 3]
```diff
@@ -44,7 +32,7 @@ export class QueryBuilder<T> {
   #query: SearchFilters
   #limit: number
   #sort?: string
-  #bookmark?: string
+  #bookmark?: string | number
```
[the remark this answers is not in the document]
Comment (mike12345567) [reply packages/backend-core/src/db/lucene.ts:35]: Discussed on call - this is correct, affects some DBs, the types were a little incorrect, making all bookmarks similar typing.

File: packages/backend-core/src/db/tests/lucene.spec.ts [hunk 4]
```diff
@@ -1,9 +1,15 @@
...
+import {
+  Database,
+  EmptyFilterOption,
+  SortOrder,
+  SortType,
+} from "@budibase/types"
+import { fullSearch, paginatedSearch, QueryBuilder } from "../lucene"
 
 const INDEX_NAME = "main"
+const TABLE_ID = ""
```
[the remark this answers is not in the document]
Comment (mike12345567) [reply packages/backend-core/src/db/tests/lucene.spec.ts]: Test broke, fixed it, made it more full proof!

**Pull request merged**